sway-lsp/src/utils/attributes.rs

```diff
@@ -1,47 +1,94 @@
 #![allow(dead_code)]
-use crate::core::token::{AstToken, Token};
-use sway_core::{language::parsed::Declaration, transform, Engines};
+use crate::core::token::{ParsedAstToken, Token, TokenAstNode, TypedAstToken};
+use sway_core::{
+    language::{parsed::Declaration, ty},
+    transform, Engines,
+};
 
 pub fn attributes_map<F>(engines: &Engines, token: &Token, mut callback: F)
 where
     F: FnMut(&transform::AttributesMap),
 {
-    match &token.parsed {
-        AstToken::Declaration(declaration) => match declaration {
-            Declaration::EnumDeclaration(decl_id) => {
-                let decl = engines.pe().get_enum(decl_id);
-                callback(&decl.attributes);
+    // Get the attributes map from the typed token if it exists.
+    // Otherwise, get the attributes map from the parsed ast node.
+    match &token.ast_node {
+        TokenAstNode::Typed(typed_token) => match typed_token {
+            TypedAstToken::TypedDeclaration(decl) => match decl {
+                ty::TyDecl::EnumDecl(ty::EnumDecl { decl_id, .. }) => {
+                    let enum_decl = engines.de().get_enum(decl_id);
+                    callback(&enum_decl.attributes);
+                }
+                ty::TyDecl::StructDecl(ty::StructDecl { decl_id, .. }) => {
+                    let struct_decl = engines.de().get_struct(decl_id);
+                    callback(&struct_decl.attributes);
+                }
+                ty::TyDecl::StorageDecl(ty::StorageDecl { decl_id, .. }) => {
+                    let storage_decl = engines.de().get_storage(decl_id);
+                    callback(&storage_decl.attributes);
+                }
+                ty::TyDecl::AbiDecl(ty::AbiDecl { decl_id, .. }) => {
+                    let abi_decl = engines.de().get_abi(decl_id);
+                    callback(&abi_decl.attributes);
+                }
```

Comment:
I think there are a few missing here:

- ty::TyDecl::FunctionDecl
- ty::TyDecl::ConstantDecl

Reply:
these are handled separately outside of the `TypedDeclaration` in this section. Good call though

Reply:
From what I remember, sometimes the function declaration tokens will come as `TypedDeclaration`s and sometimes they'll be `ty::TyDecl::FunctionDecl`. There was a bug in autocomplete where it wouldn't work for every function because I was only handling one of the decl types. This might've changed though!

Reply:
Thanks Sophie. Im going to rework how we handle this part going forward to remove this confusion.